---
title: no-children-prop
description: Prevent passing children as a prop. This rule is part of @eslint/eslint-plugin-react-features.
category: quality
severity: low
tags: ['quality', 'react']
autofix: suggestions
affects: ['readability', 'maintainability']
effort: low
---

> **Keywords:** React, children, props, composition, ESLint rule, best practices, LLM-optimized

Prevent passing `children` as a prop. This rule is part of [`@eslint/eslint-plugin-react-features`](https://www.npmjs.com/package/@eslint/eslint-plugin-react-features).

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Warning (style)                                                      |
| **Auto-Fix**   | ‚ùå No (requires restructuring)                                       |
| **Category**   | React                                                                |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration                              |
| **Best For**   | Consistent JSX patterns                                              |

## Rule Details

Passing `children` as a prop instead of nesting is less readable and can cause issues.

### Why This Matters

| Issue                     | Impact                          | Solution                  |
| ------------------------- | ------------------------------- | ------------------------- |
| üìñ **Readability**        | Less intuitive structure        | Use nested children       |
| üîÑ **Consistency**        | Mixed patterns                  | Standardize               |
| üêõ **Edge cases**         | Prop overwrite issues           | Natural nesting           |

## Examples

### ‚ùå Incorrect

```tsx
<Component children={<span>Hello</span>} />
<Component children="Hello" />
<Component children={[<span key="1">A</span>, <span key="2">B</span>]} />
```

### ‚úÖ Correct

```tsx
<Component>
  <span>Hello</span>
</Component>

<Component>Hello</Component>

<Component>
  <span key="1">A</span>
  <span key="2">B</span>
</Component>
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'react-features/no-children-prop': 'warn'
  }
}
```

## Related Rules

- [`jsx-key`](./jsx-key.md) - Keys in children lists

## Further Reading

- **[Composition vs Inheritance](https://react.dev/learn/passing-props-to-a-component)** - React docs
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Prop from variable
const propValue = computedValue;
<Component prop={propValue} /> // Computation not analyzed
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Wrapped or Aliased Functions

**Why**: Custom wrapper functions or aliased methods are not recognized by the rule.

```typescript
// ‚ùå NOT DETECTED - Custom wrapper
function myWrapper(data) {
  return internalApi(data); // Wrapper not analyzed
}
myWrapper(unsafeInput);
```

**Mitigation**: Apply this rule's principles to wrapper function implementations. Avoid aliasing security-sensitive functions.

### Imported Values

**Why**: When values come from imports, the rule cannot analyze their origin or construction.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Ensure imported values follow the same constraints. Use TypeScript for type safety.
