---
title: no-did-mount-set-state
description: "Prevent calling setState in componentDidMount. This rule is part of @eslint/eslint-plugin-react-features."
---

> **Keywords:** React, componentDidMount, setState, lifecycle, performance, ESLint rule, LLM-optimized

Prevent calling `setState` in `componentDidMount`. This rule is part of [`@eslint/eslint-plugin-react-features`](https://www.npmjs.com/package/@eslint/eslint-plugin-react-features).

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Warning (performance)                                                |
| **Auto-Fix**   | ‚ùå No (requires refactoring)                                         |
| **Category**   | React                                                                |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration                              |
| **Best For**   | React class components                                               |

## Rule Details

Calling `setState` immediately in `componentDidMount` causes an additional render cycle. Initial state should be set in the constructor or derived from props.

### Why This Matters

| Issue                         | Impact                              | Solution                     |
| ----------------------------- | ----------------------------------- | ---------------------------- |
| üîÑ **Double render**          | Performance degradation             | Set state in constructor     |
| üêõ **Wasted cycles**          | Unnecessary DOM updates             | Use derived state            |
| üîç **Flicker**                | UI may flash between states         | Initialize state properly    |

## Examples

### ‚ùå Incorrect

```tsx
class UserProfile extends React.Component {
  state = { user: null };
  
  componentDidMount() {
    // BAD: Immediate setState causes extra render
    this.setState({ loading: true });
    
    // This is also problematic
    this.setState({
      windowWidth: window.innerWidth
    });
  }
}
```

### ‚úÖ Correct

```tsx
class UserProfile extends React.Component {
  constructor(props) {
    super(props);
    // Initialize state in constructor
    this.state = {
      loading: true,
      windowWidth: typeof window !== 'undefined' ? window.innerWidth : 0
    };
  }
  
  componentDidMount() {
    // Only setState for async operations is acceptable
    fetchUser().then(user => {
      this.setState({ user, loading: false });
    });
  }
}

// Better: Use functional components with hooks
function UserProfile() {
  const [user, setUser] = useState(null);
  const [loading, setLoading] = useState(true);
  
  useEffect(() => {
    fetchUser().then(user => {
      setUser(user);
      setLoading(false);
    });
  }, []);
}
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'react-features/no-did-mount-set-state': 'warn'
  }
}
```

## Related Rules

- [`no-did-update-set-state`](./no-did-update-set-state.md) - Similar for componentDidUpdate
- [`react-class-to-hooks`](./react-class-to-hooks.md) - Migrate to hooks

## Further Reading

- **[componentDidMount](https://react.dev/reference/react/Component#componentdidmount)** - React docs
- **[State Initialization](https://react.dev/reference/react/Component#constructor)** - Constructor patterns
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Prop from variable
const propValue = computedValue;
<Component prop={propValue} /> // Computation not analyzed
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Wrapped or Aliased Functions

**Why**: Custom wrapper functions or aliased methods are not recognized by the rule.

```typescript
// ‚ùå NOT DETECTED - Custom wrapper
function myWrapper(data) {
  return internalApi(data); // Wrapper not analyzed
}
myWrapper(unsafeInput);
```

**Mitigation**: Apply this rule's principles to wrapper function implementations. Avoid aliasing security-sensitive functions.

### Imported Values

**Why**: When values come from imports, the rule cannot analyze their origin or construction.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Ensure imported values follow the same constraints. Use TypeScript for type safety.
