---
title: no-typos
description: "Detect common typos in React component properties and lifecycle methods. This rule is part of @eslint/eslint-plugin-react-features."
---

> **Keywords:** React, typos, lifecycle, static properties, ESLint rule, autocorrect, LLM-optimized

Detect common typos in React component properties and lifecycle methods. This rule is part of [`@eslint/eslint-plugin-react-features`](https://www.npmjs.com/package/@eslint/eslint-plugin-react-features).

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Error (correctness)                                                  |
| **Auto-Fix**   | üí° Suggests corrections                                              |
| **Category**   | React                                                                |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration                              |
| **Best For**   | All React projects                                                   |

## Rule Details

Detects misspellings in React lifecycle methods and static properties.

### Common Typos Detected

| Typo                  | Correct                    |
| --------------------- | -------------------------- |
| `componentDidUpate`   | `componentDidUpdate`       |
| `componentWillUnmout` | `componentWillUnmount`     |
| `getDerivedStateFromProp` | `getDerivedStateFromProps` |
| `propTypes` (wrong case) | `propTypes`             |
| `defaultProps` (wrong) | `defaultProps`            |

## Examples

### ‚ùå Incorrect

```tsx
class MyComponent extends React.Component {
  // Typo in lifecycle
  componentDidUpate() {}  // Should be componentDidUpdate
  componentWillUnmout() {}  // Should be componentWillUnmount
  
  // Typo in static property
  static PropTypes = {};  // Should be propTypes (lowercase p)
}
```

### ‚úÖ Correct

```tsx
class MyComponent extends React.Component {
  componentDidUpdate() {}
  componentWillUnmount() {}
  
  static propTypes = {};
  static defaultProps = {};
}
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'react-features/no-typos': 'error'
  }
}
```

## Related Rules

- [`display-name`](./display-name.md) - Component naming
- [`react-class-to-hooks`](./react-class-to-hooks.md) - Migration to hooks

## Further Reading

- **[React Component Lifecycle](https://react.dev/reference/react/Component)** - React docs
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Prop from variable
const propValue = computedValue;
<Component prop={propValue} /> // Computation not analyzed
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Wrapped or Aliased Functions

**Why**: Custom wrapper functions or aliased methods are not recognized by the rule.

```typescript
// ‚ùå NOT DETECTED - Custom wrapper
function myWrapper(data) {
  return internalApi(data); // Wrapper not analyzed
}
myWrapper(unsafeInput);
```

**Mitigation**: Apply this rule's principles to wrapper function implementations. Avoid aliasing security-sensitive functions.

### Imported Values

**Why**: When values come from imports, the rule cannot analyze their origin or construction.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Ensure imported values follow the same constraints. Use TypeScript for type safety.
