---
title: require-render-return
description: "Require return statement in render methods. This rule is part of @eslint/eslint-plugin-react-features."
---

> **Keywords:** React, render, return, class component, ESLint rule, correctness, LLM-optimized

Require `return` statement in render methods. This rule is part of [`@eslint/eslint-plugin-react-features`](https://www.npmjs.com/package/@eslint/eslint-plugin-react-features).

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Error (correctness)                                                  |
| **Auto-Fix**   | ‚ùå No                                                                |
| **Category**   | React                                                                |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration                              |
| **Best For**   | React class components                                               |

## Rule Details

The `render()` method must return a React element, `null`, or nothing (`undefined` implicitly returns `null`).

### Why This Matters

| Issue                     | Impact                          | Solution                  |
| ------------------------- | ------------------------------- | ------------------------- |
| üêõ **Silent failure**     | Component renders nothing       | Add return statement      |
| ‚ùì **Unclear intent**     | Is missing return intentional?  | Explicit return           |
| üß™ **Testing**            | Hard to catch in tests          | Static analysis           |

## Examples

### ‚ùå Incorrect

```tsx
class MyComponent extends React.Component {
  render() {
    <div>Hello</div>;  // Missing return!
  }
}

class AnotherComponent extends React.Component {
  render() {
    if (this.props.show) {
      return <div>Content</div>;
    }
    // Missing return for else case
  }
}
```

### ‚úÖ Correct

```tsx
class MyComponent extends React.Component {
  render() {
    return <div>Hello</div>;
  }
}

class AnotherComponent extends React.Component {
  render() {
    if (this.props.show) {
      return <div>Content</div>;
    }
    return null;  // Explicit return
  }
}
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'react-features/require-render-return': 'error'
  }
}
```

## Related Rules

- [`display-name`](./display-name.md) - Component naming
- [`prefer-stateless-function`](./prefer-stateless-function.md) - Function components

## Further Reading

- **[Render Method](https://react.dev/reference/react/Component#render)** - React docs
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Prop from variable
const propValue = computedValue;
<Component prop={propValue} /> // Computation not analyzed
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Cross-Module Data Flow

**Why**: ESLint rules analyze one file at a time. Values imported from other modules cannot be traced.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Apply the same rule to imported modules. Use module boundaries and explicit exports.
