# jsx-key

> **Keywords:** React, JSX, key prop, reconciliation, lists, map, iteration, ESLint rule, performance, LLM-optimized

Detect missing or problematic React keys that could break reconciliation. This rule is part of [`@eslint/eslint-plugin-react-features`](https://www.npmjs.com/package/@eslint/eslint-plugin-react-features) and provides LLM-optimized error messages with suggestions.

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Error (correctness)                                                  |
| **Auto-Fix**   | üí° Suggests fixes                                                    |
| **Category**   | React                                                                |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration                              |
| **Best For**   | All React/JSX projects                                               |

## Rule Details

```mermaid
%%{init: {
  'theme': 'base',
  'themeVariables': {
    'primaryColor': '#f8fafc',
    'primaryTextColor': '#1e293b',
    'primaryBorderColor': '#334155',
    'lineColor': '#475569',
    'c0': '#f8fafc',
    'c1': '#f1f5f9',
    'c2': '#e2e8f0',
    'c3': '#cbd5e1'
  }
}}%%
flowchart TD
    A[üîç JSX in Iterator] --> B{Has key prop?}
    B -->|‚ùå No| C[‚ùå Report: Missing Key]
    B -->|‚úÖ Yes| D{Key Quality?}
    
    D -->|üî¢ Index| E[‚ö†Ô∏è Warn: Unstable Key]
    D -->|‚úÖ Stable ID| F[‚úÖ Pass]
    D -->|üîÅ Duplicate| G[‚ùå Report: Duplicate Key]
    
    classDef startNode fill:#f0fdf4,stroke:#16a34a,stroke-width:2px,color:#1f2937
    classDef errorNode fill:#fef2f2,stroke:#dc2626,stroke-width:2px,color:#1f2937
    classDef warnNode fill:#fef3c7,stroke:#f59e0b,stroke-width:2px,color:#1f2937
    classDef processNode fill:#eff6ff,stroke:#2563eb,stroke-width:2px,color:#1f2937
    
    class A startNode
    class C,G errorNode
    class E warnNode
    class F processNode
```

### Why This Matters

| Issue                     | Impact                          | Solution                  |
| ------------------------- | ------------------------------- | ------------------------- |
| üîÑ **Reconciliation**     | React can't track elements      | Add unique keys           |
| ‚ö° **Performance**        | Unnecessary re-renders          | Stable keys               |
| üêõ **State Bugs**         | Wrong component gets state      | Use item IDs, not indexes |
| üé® **Animation Issues**   | Elements animate incorrectly    | Consistent key identity   |

## Configuration

| Option             | Type      | Default | Description                              |
| ------------------ | --------- | ------- | ---------------------------------------- |
| `warnUnstableKeys` | `boolean` | `true`  | Warn about potentially unstable keys     |

## Examples

### ‚ùå Incorrect

```tsx
// Missing key
function UserList({ users }) {
  return (
    <ul>
      {users.map(user => (
        <li>{user.name}</li>  // ‚ùå Missing key
      ))}
    </ul>
  );
}

// Using index as key (unstable)
function ItemList({ items }) {
  return (
    <ul>
      {items.map((item, index) => (
        <li key={index}>{item.name}</li>  // ‚ö†Ô∏è Unstable key
      ))}
    </ul>
  );
}
```

### ‚úÖ Correct

```tsx
// Using unique ID as key
function UserList({ users }) {
  return (
    <ul>
      {users.map(user => (
        <li key={user.id}>{user.name}</li>  // ‚úÖ Stable unique key
      ))}
    </ul>
  );
}

// Using compound key when no ID available
function ItemList({ items }) {
  return (
    <ul>
      {items.map(item => (
        <li key={`${item.category}-${item.name}`}>
          {item.name}
        </li>  // ‚úÖ Unique compound key
      ))}
    </ul>
  );
}
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'react-features/jsx-key': 'error'
  }
}
```

### Disable Unstable Key Warnings

```javascript
{
  rules: {
    'react-features/jsx-key': ['error', {
      warnUnstableKeys: false
    }]
  }
}
```

## Key Best Practices

### Choosing Keys

| Source              | Quality    | Example                              |
| ------------------- | ---------- | ------------------------------------ |
| Database ID         | ‚úÖ Best    | `key={user.id}`                      |
| Unique field        | ‚úÖ Good    | `key={item.slug}`                    |
| Compound unique     | ‚úÖ Good    | `key={\`${cat}-${name}\`}`           |
| UUID/nanoid         | ‚ö†Ô∏è OK      | `key={generateId()}`                 |
| Array index         | ‚ùå Avoid   | `key={index}`                        |
| Random number       | ‚ùå Never   | `key={Math.random()}`                |

### When Index Keys Are Acceptable

```tsx
// ‚úÖ Static list that never reorders/filters
const WEEKDAYS = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri'];
<ul>
  {WEEKDAYS.map((day, i) => <li key={i}>{day}</li>)}
</ul>

// ‚ùå Dynamic list - DON'T use index
{users.map((user, i) => <UserCard key={i} user={user} />)}  // Bad!
{users.map(user => <UserCard key={user.id} user={user} />)} // Good!
```

## Common Patterns

### Fragment Keys

```tsx
// When using fragments in lists, key goes on Fragment
function DataList({ data }) {
  return (
    <>
      {data.map(item => (
        <React.Fragment key={item.id}>
          <dt>{item.term}</dt>
          <dd>{item.definition}</dd>
        </React.Fragment>
      ))}
    </>
  );
}
```

### Nested Lists

```tsx
function NestedList({ categories }) {
  return categories.map(category => (
    <div key={category.id}>
      <h2>{category.name}</h2>
      <ul>
        {category.items.map(item => (
          <li key={item.id}>{item.name}</li>  // Separate key namespace
        ))}
      </ul>
    </div>
  ));
}
```

## When Not To Use

| Scenario                    | Recommendation                              |
| --------------------------- | ------------------------------------------- |
| üß™ **Testing/Prototypes**   | Consider allowing index keys temporarily    |
| üìä **Static content**       | Index keys may be acceptable                |
| üîÑ **No reordering**        | Index keys work for append-only lists       |

## Comparison with Alternatives

| Feature              | jsx-key             | eslint-plugin-react | jsx-ally           |
| -------------------- | ------------------- | ------------------- | ------------------ |
| **Missing key**      | ‚úÖ Yes              | ‚úÖ Yes              | ‚ùå No              |
| **Unstable keys**    | ‚úÖ Configurable     | ‚ö†Ô∏è Limited          | ‚ùå No              |
| **LLM-Optimized**    | ‚úÖ Yes              | ‚ùå No               | ‚ùå No              |
| **ESLint MCP**       | ‚úÖ Optimized        | ‚ùå No               | ‚ùå No              |
| **Suggestions**      | ‚úÖ Yes              | ‚ö†Ô∏è Limited          | ‚ùå No              |

## Related Rules

- [`react-no-inline-functions`](./react-no-inline-functions.md) - Performance optimization
- [`react-render-optimization`](./react-render-optimization.md) - Render performance

## Further Reading

- **[React Keys Documentation](https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key)** - Official React docs
- **[Why React Keys Matter](https://kentcdodds.com/blog/understanding-reacts-key-prop)** - Kent C. Dodds article
- **[Reconciliation](https://react.dev/learn/preserving-and-resetting-state)** - How React updates the DOM
- **[ESLint MCP Setup](https://eslint.org/docs/latest/use/mcp)** - Enable AI assistant integration
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Prop from variable
const propValue = computedValue;
<Component prop={propValue} /> // Computation not analyzed
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Imported Values

**Why**: When values come from imports, the rule cannot analyze their origin or construction.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Ensure imported values follow the same constraints. Use TypeScript for type safety.



