# no-children-prop

> **Keywords:** React, children, props, composition, ESLint rule, best practices, LLM-optimized

Prevent passing `children` as a prop. This rule is part of [`@eslint/eslint-plugin-react-features`](https://www.npmjs.com/package/@eslint/eslint-plugin-react-features).

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Warning (style)                                                      |
| **Auto-Fix**   | âŒ No (requires restructuring)                                       |
| **Category**   | React                                                                |
| **ESLint MCP** | âœ… Optimized for ESLint MCP integration                              |
| **Best For**   | Consistent JSX patterns                                              |

## Rule Details

Passing `children` as a prop instead of nesting is less readable and can cause issues.

### Why This Matters

| Issue                     | Impact                          | Solution                  |
| ------------------------- | ------------------------------- | ------------------------- |
| ğŸ“– **Readability**        | Less intuitive structure        | Use nested children       |
| ğŸ”„ **Consistency**        | Mixed patterns                  | Standardize               |
| ğŸ› **Edge cases**         | Prop overwrite issues           | Natural nesting           |

## Examples

### âŒ Incorrect

```tsx
<Component children={<span>Hello</span>} />
<Component children="Hello" />
<Component children={[<span key="1">A</span>, <span key="2">B</span>]} />
```

### âœ… Correct

```tsx
<Component>
  <span>Hello</span>
</Component>

<Component>Hello</Component>

<Component>
  <span key="1">A</span>
  <span key="2">B</span>
</Component>
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'eslint-plugin/no-children-prop': 'warn'
  }
}
```

## Related Rules

- [`jsx-key`](./jsx-key.md) - Keys in children lists

## Further Reading

- **[Composition vs Inheritance](https://react.dev/learn/passing-props-to-a-component)** - React docs

