---
title: alt-text
description: Require alt text for images and other media elements. This rule is part of @interlace/eslint-plugin-react-a11y and provides LLM-optimized error messages.
category: quality
severity: low
tags: ['quality', 'react', 'accessibility', 'a11y']
autofix: suggestions
affects: ['readability', 'maintainability']
effort: low
---

> **Keywords:** React, accessibility, a11y, alt text, images, ESLint rule, LLM-optimized

Require alt text for images and other media elements. This rule is part of [`@interlace/eslint-plugin-react-a11y`](https://www.npmjs.com/package/@interlace/eslint-plugin-react-a11y).

## Quick Summary

| Aspect         | Details                                 |
| -------------- | --------------------------------------- |
| **Severity**   | Error (accessibility)                   |
| **Auto-Fix**   | ❌ No auto-fix                          |
| **Category**   | Accessibility                           |
| **ESLint MCP** | ✅ Optimized for ESLint MCP integration |

## Rule Details

Images must have alternative text for screen readers. This is required by WCAG 2.0 Success Criterion 1.1.1.

## Examples

### ❌ Incorrect

```jsx
<img src="photo.jpg" />

<img src="photo.jpg" alt="" />

<Area href="/link" />
```

### ✅ Correct

```jsx
<img src="photo.jpg" alt="A sunset over the mountains" />

{/* Decorative images can use empty alt with role */}
<img src="decorative.svg" alt="" role="presentation" />

<Area href="/link" alt="Navigate to dashboard" />

<input type="image" alt="Submit form" />
```

## Configuration

```javascript
{
  rules: {
    'react-a11y/alt-text': 'error'
  }
}
```

## Related Rules

- [`img-redundant-alt`](./img-redundant-alt.mdx) - Prevents redundant words in alt text
