---
title: lang
description: Require valid lang attribute value. This rule is part of eslint-plugin-react-a11y and provides LLM-optimized error messages with fix suggestions.
---

> **Keywords:** lang, accessibility, ESLint rule, WCAG, a11y, React accessibility

Require valid lang attribute value. This rule is part of [`eslint-plugin-react-a11y`](https://www.npmjs.com/package/eslint-plugin-react-a11y) and provides LLM-optimized error messages with fix suggestions.

## Quick Summary

| Aspect | Details |
|--------|---------|
| **WCAG Criterion** | 3.1.1 Language of Page |
| **Severity** | Error/Warning (accessibility) |
| **Auto-Fix** | üí° Suggestions available |
| **Category** | Accessibility |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration |

## Rule Details

This rule helps ensure WCAG 3.1.1 compliance by enforcing: Require valid lang attribute value

### Why This Matters

| Issue | Impact | Standard |
|-------|--------|----------|
| ‚ôø **Accessibility** | Screen reader and assistive technology users affected | WCAG 3.1.1 |
| ‚öñÔ∏è **Legal** | ADA/Section 508 compliance risk | Legal Requirement |
| üîç **SEO** | Search engines prefer accessible sites | Best Practice |

## Examples

### ‚ùå Incorrect

```tsx
// Violation of lang
// See rule source for specific examples
```

### ‚úÖ Correct

```tsx
// Compliant with lang
// See rule source for specific examples
```

## Configuration

```javascript
// eslint.config.js
{
  rules: {
    'react-a11y/lang': 'error'
  }
}
```

## WCAG 2.1 Compliance

This rule helps satisfy:
- **3.1.1 Language of Page**: Require valid lang attribute value

## Related Rules

- See [RULES.md](../RULES.md) for all accessibility rules

## Further Reading

- **[WCAG 3.1.1](https://www.w3.org/WAI/WCAG21/Understanding/)** - WCAG guidelines
- **[WebAIM](https://webaim.org/)** - Accessibility resources
- **[MDN Accessibility](https://developer.mozilla.org/en-US/docs/Web/Accessibility)** - MDN documentation

## Version

This rule is available in `eslint-plugin-react-a11y` v1.0.0+
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Prop from variable
const propValue = computedValue;
<Component prop={propValue} /> // Computation not analyzed
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Imported Values

**Why**: When values come from imports, the rule cannot analyze their origin or construction.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Ensure imported values follow the same constraints. Use TypeScript for type safety.
