---
title: no-aria-hidden-on-focusable
description: Prevent aria-hidden on focusable elements. This rule is part of eslint-plugin-react-a11y and provides LLM-optimized error messages with fix suggestions.
---

> **Keywords:** no-aria-hidden-on-focusable, accessibility, ESLint rule, WCAG, a11y, React accessibility

Prevent aria-hidden on focusable elements. This rule is part of [`eslint-plugin-react-a11y`](https://www.npmjs.com/package/eslint-plugin-react-a11y) and provides LLM-optimized error messages with fix suggestions.

## Quick Summary

| Aspect | Details |
|--------|---------|
| **WCAG Criterion** | 4.1.2 Name, Role, Value |
| **Severity** | Error/Warning (accessibility) |
| **Auto-Fix** | üí° Suggestions available |
| **Category** | Accessibility |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration |

## Rule Details

This rule helps ensure WCAG 4.1.2 compliance by enforcing: Prevent aria-hidden on focusable elements

### Why This Matters

| Issue | Impact | Standard |
|-------|--------|----------|
| ‚ôø **Accessibility** | Screen reader and assistive technology users affected | WCAG 4.1.2 |
| ‚öñÔ∏è **Legal** | ADA/Section 508 compliance risk | Legal Requirement |
| üîç **SEO** | Search engines prefer accessible sites | Best Practice |

## Examples

### ‚ùå Incorrect

```tsx
// Violation of no-aria-hidden-on-focusable
// See rule source for specific examples
```

### ‚úÖ Correct

```tsx
// Compliant with no-aria-hidden-on-focusable
// See rule source for specific examples
```

## Configuration

```javascript
// eslint.config.js
{
  rules: {
    'react-a11y/no-aria-hidden-on-focusable': 'error'
  }
}
```

## WCAG 2.1 Compliance

This rule helps satisfy:
- **4.1.2 Name, Role, Value**: Prevent aria-hidden on focusable elements

## Related Rules

- See [RULES.md](../RULES.md) for all accessibility rules

## Further Reading

- **[WCAG 4.1.2](https://www.w3.org/WAI/WCAG21/Understanding/)** - WCAG guidelines
- **[WebAIM](https://webaim.org/)** - Accessibility resources
- **[MDN Accessibility](https://developer.mozilla.org/en-US/docs/Web/Accessibility)** - MDN documentation

## Version

This rule is available in `eslint-plugin-react-a11y` v1.0.0+
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Prop from variable
const propValue = computedValue;
<Component prop={propValue} /> // Computation not analyzed
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Wrapped or Aliased Functions

**Why**: Custom wrapper functions or aliased methods are not recognized by the rule.

```typescript
// ‚ùå NOT DETECTED - Custom wrapper
function myWrapper(data) {
  return internalApi(data); // Wrapper not analyzed
}
myWrapper(unsafeInput);
```

**Mitigation**: Apply this rule's principles to wrapper function implementations. Avoid aliasing security-sensitive functions.

### Imported Values

**Why**: When values come from imports, the rule cannot analyze their origin or construction.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Ensure imported values follow the same constraints. Use TypeScript for type safety.
