---
title: no-instanceof-array
description: Require Array.isArray() instead of instanceof Array. This rule is part of @eslint/eslint-plugin-architecture.
category: quality
severity: low
tags: ['quality', 'modernization', 'es6+']
autofix: suggestions
affects: ['readability', 'maintainability']
effort: low
---

> **Keywords:** instanceof, Array.isArray, type checking, ESLint rule, cross-realm, LLM-optimized

Require `Array.isArray()` instead of `instanceof Array`. This rule is part of [`@eslint/eslint-plugin-architecture`](https://www.npmjs.com/package/@eslint/eslint-plugin-architecture).

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Warning (correctness)                                                |
| **Auto-Fix**   | ‚úÖ Yes (converts to Array.isArray)                                   |
| **Category**   | Architecture                                                         |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration                              |
| **Best For**   | Cross-realm code, iframes, Web Workers                               |

## Rule Details

`instanceof Array` fails for arrays from different JavaScript realms (iframes, Web Workers).

### Why This Matters

| Issue                     | Impact                          | Solution                  |
| ------------------------- | ------------------------------- | ------------------------- |
| üñºÔ∏è **Cross-realm**        | False negatives                 | Array.isArray()           |
| üîß **Web Workers**        | Type checks fail                | Static method             |
| üì¶ **Library code**       | Unexpected behavior             | Universal check           |

## Examples

### ‚ùå Incorrect

```typescript
if (value instanceof Array) {
  // May fail for arrays from iframes/workers
}

function isArray(val: unknown) {
  return val instanceof Array;  // Unreliable
}
```

### ‚úÖ Correct

```typescript
if (Array.isArray(value)) {
  // Works across realms
}

function isArray(val: unknown): val is unknown[] {
  return Array.isArray(val);  // Reliable
}
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'architecture/no-instanceof-array': 'warn'
  }
}
```

## Related Rules

- [`no-unsafe-type-narrowing`](./no-unsafe-type-narrowing.md) - Type narrowing safety

## Further Reading

- **[Array.isArray() - MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray)** - MDN reference
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Value from variable
const value = externalSource();
processValue(value); // Variable origin not tracked
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Wrapped or Aliased Functions

**Why**: Custom wrapper functions or aliased methods are not recognized by the rule.

```typescript
// ‚ùå NOT DETECTED - Custom wrapper
function myWrapper(data) {
  return internalApi(data); // Wrapper not analyzed
}
myWrapper(unsafeInput);
```

**Mitigation**: Apply this rule's principles to wrapper function implementations. Avoid aliasing security-sensitive functions.

### Imported Values

**Why**: When values come from imports, the rule cannot analyze their origin or construction.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Ensure imported values follow the same constraints. Use TypeScript for type safety.
