---
title: prefer-at
description: Prefer using Array.at() for accessing elements, especially with negative indices. This rule is part of @eslint/eslint-plugin-architecture.
category: quality
severity: low
tags: ['quality', 'modernization', 'es6+']
autofix: suggestions
affects: ['readability', 'maintainability']
effort: low
---

> **Keywords:** Array.at(), negative index, last element, ESLint rule, ES2022, auto-fix, LLM-optimized

Prefer using `Array.at()` for accessing elements, especially with negative indices. This rule is part of [`@eslint/eslint-plugin-architecture`](https://www.npmjs.com/package/@eslint/eslint-plugin-architecture).

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Warning (modern JavaScript)                                          |
| **Auto-Fix**   | ‚úÖ Yes (converts to .at())                                           |
| **Category**   | Architecture                                                         |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration                              |
| **Best For**   | ES2022+ codebases, cleaner array access                              |

## Rule Details

`Array.at()` provides a cleaner way to access array elements, especially the last element or elements from the end.

### Why This Matters

| Issue                     | Impact                          | Solution                  |
| ------------------------- | ------------------------------- | ------------------------- |
| üìñ **Readability**        | `arr[arr.length - 1]` is verbose| Use `arr.at(-1)`          |
| üêõ **Off-by-one errors**  | Easy to make mistakes           | Negative indices          |
| üîÑ **Consistency**        | Multiple patterns in codebase   | Standardize on .at()      |

## Examples

### ‚ùå Incorrect

```typescript
// Accessing last element
const last = array[array.length - 1];

// Second to last
const secondLast = array[array.length - 2];

// Dynamic negative access
const item = array[array.length - offset];
```

### ‚úÖ Correct

```typescript
// Clean last element access
const last = array.at(-1);

// Second to last
const secondLast = array.at(-2);

// Dynamic negative access
const item = array.at(-offset);

// Also works with strings
const lastChar = string.at(-1);
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'architecture/prefer-at': 'warn'
  }
}
```

## Related Rules

- [`prefer-node-protocol`](./prefer-node-protocol.md) - Modern Node.js imports

## Further Reading

- **[Array.at() - MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at)** - MDN reference
- **[ES2022 Features](https://tc39.es/ecma262/)** - ECMAScript specification
## Known False Negatives

The following patterns are **not detected** due to static analysis limitations:

### Dynamic Variable References

**Why**: Static analysis cannot trace values stored in variables or passed through function parameters.

```typescript
// ‚ùå NOT DETECTED - Value from variable
const value = externalSource();
processValue(value); // Variable origin not tracked
```

**Mitigation**: Implement runtime validation and review code manually. Consider using TypeScript branded types for validated inputs.

### Imported Values

**Why**: When values come from imports, the rule cannot analyze their origin or construction.

```typescript
// ‚ùå NOT DETECTED - Value from import
import { getValue } from './helpers';
processValue(getValue()); // Cross-file not tracked
```

**Mitigation**: Ensure imported values follow the same constraints. Use TypeScript for type safety.
