# no-instanceof-array

> **Keywords:** instanceof, Array.isArray, type checking, ESLint rule, cross-realm, LLM-optimized

Require `Array.isArray()` instead of `instanceof Array`. This rule is part of [`@eslint/eslint-plugin-architecture`](https://www.npmjs.com/package/@eslint/eslint-plugin-architecture).

## Quick Summary

| Aspect         | Details                                                              |
| -------------- | -------------------------------------------------------------------- |
| **Severity**   | Warning (correctness)                                                |
| **Auto-Fix**   | ‚úÖ Yes (converts to Array.isArray)                                   |
| **Category**   | Architecture                                                         |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration                              |
| **Best For**   | Cross-realm code, iframes, Web Workers                               |

## Rule Details

`instanceof Array` fails for arrays from different JavaScript realms (iframes, Web Workers).

### Why This Matters

| Issue                     | Impact                          | Solution                  |
| ------------------------- | ------------------------------- | ------------------------- |
| üñºÔ∏è **Cross-realm**        | False negatives                 | Array.isArray()           |
| üîß **Web Workers**        | Type checks fail                | Static method             |
| üì¶ **Library code**       | Unexpected behavior             | Universal check           |

## Examples

### ‚ùå Incorrect

```typescript
if (value instanceof Array) {
  // May fail for arrays from iframes/workers
}

function isArray(val: unknown) {
  return val instanceof Array;  // Unreliable
}
```

### ‚úÖ Correct

```typescript
if (Array.isArray(value)) {
  // Works across realms
}

function isArray(val: unknown): val is unknown[] {
  return Array.isArray(val);  // Reliable
}
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'architecture/no-instanceof-array': 'warn'
  }
}
```

## Related Rules

- [`no-unsafe-type-narrowing`](./no-unsafe-type-narrowing.md) - Type narrowing safety

## Further Reading

- **[Array.isArray() - MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray)** - MDN reference

