# export

ðŸ’¼ This rule is enabled in the following configs: `recommended`, `typescript`.
ðŸ’¡ This rule is automatically fixable by the `--fix` CLI option.

<!-- end auto-generated rule header -->

Forbid any invalid exports, i.e. re-export of the same name

## Rule Details

This rule aims to prevent issues related to export.

## Options

```json
[]
```

## Implementation

- [Source Code](https://github.com/import-js/eslint-plugin-import-next/blob/main/src/rules/export.ts)
- [Test Cases](https://github.com/import-js/eslint-plugin-import-next/blob/main/src/tests/export.test.ts)

## OWASP Foundation

- **Category**: A00:2021 - General Security
