---
title: error-message
description: Enforce meaningful error messages with context for better debugging. This rule is part of @interlace/eslint-plugin-maintainability and provides LLM-optimized error messages.
category: quality
severity: medium
tags: ['quality', 'maintainability', 'error-handling']
autofix: suggestions
affects: ['readability', 'maintainability']
effort: low
---

> **Keywords:** error, message, debugging, context, ESLint rule, code quality, maintainability, LLM-optimized

Enforce meaningful error messages with context for better debugging. This rule is part of [`@interlace/eslint-plugin-maintainability`](https://www.npmjs.com/package/@interlace/eslint-plugin-maintainability) and provides LLM-optimized error messages with suggestions.

## Quick Summary

| Aspect         | Details                                 |
| -------------- | --------------------------------------- |
| **Severity**   | Warning (code quality)                  |
| **Auto-Fix**   | üí° Suggests fixes                       |
| **Category**   | Error Handling                          |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration |
| **Best For**   | All projects prioritizing debuggability |

## Rule Details

This rule ensures that error messages provide meaningful context to help with debugging and troubleshooting.

### Why This Matters

| Issue                | Impact                     | Solution                 |
| -------------------- | -------------------------- | ------------------------ |
| üìñ **Debuggability** | Hard to find error source  | Add context to messages  |
| üêõ **Bug Detection** | Generic errors hide issues | Include specific details |
| üîÑ **Maintenance**   | Time wasted investigating  | Clear error descriptions |
| üëÄ **Monitoring**    | Harder to track in logs    | Structured error info    |

## Examples

### ‚ùå Incorrect

```typescript
// Generic error message
throw new Error('Error');

// Empty error message
throw new Error('');

// Too vague
throw new Error('Something went wrong');

// No context about what failed
throw new Error('Invalid input');
```

### ‚úÖ Correct

```typescript
// Specific context about what failed
throw new Error(`Failed to fetch user ${userId}: ${response.status}`);

// Clear description of the issue
throw new Error(`Invalid email format: expected '@' symbol in "${email}"`);

// Actionable error message
throw new Error(
  `Configuration file not found at ${configPath}. Run 'init' first.`,
);

// Include relevant values
throw new Error(`Payment amount ${amount} exceeds maximum ${MAX_AMOUNT}`);
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'maintainability/error-message': 'warn'
  }
}
```

### Strict Mode

```javascript
{
  rules: {
    'maintainability/error-message': 'error'
  }
}
```

## Related Rules

- [`no-silent-errors`](./no-silent-errors.mdx) - Prevent ignoring errors
- [`no-missing-error-context`](./no-missing-error-context.mdx) - Similar contextual requirements
- [`no-unhandled-promise`](./no-unhandled-promise.mdx) - Async error handling

## Further Reading

- **[Error Handling Best Practices](https://nodejs.org/en/docs/guides/error-handling)** - Node.js error handling guide
- **[ESLint MCP Setup](https://eslint.org/docs/latest/use/mcp)** - Enable AI assistant integration
