---
title: no-nested-ternary
description: Prevent nested ternary expressions for better readability
category: quality
severity: medium
tags: ['quality', 'readability', 'complexity', 'refactoring']
autofix: suggestions
affects: ['readability', 'maintainability']
effort: low
---

> **Keywords:** ternary, conditional, nested, readability, ESLint rule, code quality, refactoring

Prevent nested ternary expressions for better readability. This rule is part of [`@interlace/eslint-plugin-maintainability`](https://www.npmjs.com/package/@interlace/eslint-plugin-maintainability).

## Quick Summary

| Aspect         | Details                                               |
| -------------- | ----------------------------------------------------- |
| **Severity**   | Warning (code quality)                                |
| **Category**   | Quality                                               |
| **Auto-Fix**   | üí° Suggestions                                        |
| **Best For**   | Projects prioritizing readability and maintainability |
| **ESLint MCP** | ‚úÖ Optimized for ESLint MCP integration               |

## Why This Matters

| Issue                  | Impact                | Solution               |
| ---------------------- | --------------------- | ---------------------- |
| üìñ **Readability**     | Hard to follow logic  | Use if-else or extract |
| üêõ **Bug Prevention**  | Easy to make mistakes | Simpler constructs     |
| üîÑ **Maintainability** | Difficult to modify   | Break into steps       |
| üëÄ **Code Review**     | Harder to review      | Clearer structure      |

## Rule Details

```mermaid
%%{init: {
  'theme': 'base',
  'themeVariables': {
    'primaryColor': '#f8fafc',
    'primaryTextColor': '#1e293b',
    'primaryBorderColor': '#334155',
    'lineColor': '#475569'
  }
}}%%
flowchart TD
    A[üîç Detect Ternary] --> B{Contains Nested?}
    B -->|‚ùå No| C[‚úÖ Pass]
    B -->|‚úÖ Yes| D{In Allowed Context?}

    D -->|‚úÖ Yes| C
    D -->|‚ùå No| E[‚ö†Ô∏è Report: Extract or use if-else]
```

## Examples

### ‚ùå Incorrect

```typescript
// Nested ternary in consequent
const result = condition1 ? (condition2 ? value1 : value2) : value3;

// Multiple levels of nesting
const color =
  size === 'large'
    ? 'red'
    : size === 'medium'
      ? 'blue'
      : size === 'small'
        ? 'green'
        : 'gray';
```

### ‚úÖ Correct

```typescript
// Single ternary
const result = condition ? value1 : value2;

// Use if-else for complex logic
let status: string;
if (isActive) {
  status = 'active';
} else if (isPending) {
  status = 'pending';
} else {
  status = 'inactive';
}

// Use object lookup
const colorMap: Record<string, string> = {
  large: 'red',
  medium: 'blue',
  small: 'green',
};
const color = colorMap[size] ?? 'gray';
```

## Configuration

```javascript
{
  "maintainability/no-nested-ternary": ["warn", {
    "allow": []
  }]
}
```

| Option  | Type       | Default | Description                             |
| ------- | ---------- | ------- | --------------------------------------- |
| `allow` | `string[]` | `[]`    | Contexts where nested ternaries allowed |

## Refactoring Patterns

### Pattern 1: Extract to Variable

```typescript
// Before
const message = user.isAdmin
  ? user.isActive
    ? 'Active Admin'
    : 'Inactive Admin'
  : 'Regular User';

// After
const adminStatus = user.isActive ? 'Active Admin' : 'Inactive Admin';
const message = user.isAdmin ? adminStatus : 'Regular User';
```

### Pattern 2: Use Object Lookup

```typescript
// Before
const icon = status === 'success' ? '‚úì' : status === 'error' ? '‚úó' : '‚Ä¢';

// After
const statusIcons: Record<string, string> = {
  success: '‚úì',
  error: '‚úó',
};
const icon = statusIcons[status] ?? '‚Ä¢';
```

## When Not To Use It

- Very simple cases where nested ternary is still readable
- JSX conditionals with proper formatting
- TypeScript type narrowing patterns

## Related Rules

- [`cognitive-complexity`](./cognitive-complexity.mdx) - Overall code complexity
- [`no-lonely-if`](./no-lonely-if.mdx) - Simplify if statements

## Further Reading

- [ESLint no-nested-ternary](https://eslint.org/docs/latest/rules/no-nested-ternary)
- [Clean Code](https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882)
