---
title: no-unhandled-promise
description: Ensure all promises have proper error handling. This rule is part of @interlace/eslint-plugin-maintainability and provides LLM-optimized error messages.
category: quality
severity: low
tags: ['quality', 'maintainability', 'async']
autofix: suggestions
affects: ['readability', 'maintainability']
effort: low
---

> **Keywords:** promise, async, await, error handling, ESLint rule, maintainability, LLM-optimized

Ensure all promises have proper error handling. This rule is part of [`@interlace/eslint-plugin-maintainability`](https://www.npmjs.com/package/@interlace/eslint-plugin-maintainability) and provides LLM-optimized error messages with suggestions.

## Quick Summary

| Aspect         | Details                                 |
| -------------- | --------------------------------------- |
| **Severity**   | Error (critical)                        |
| **Auto-Fix**   | ðŸ’¡ Suggests fixes                       |
| **Category**   | Error Handling                          |
| **ESLint MCP** | âœ… Optimized for ESLint MCP integration |
| **Best For**   | All async JavaScript/TypeScript code    |

## Rule Details

This rule detects promises that are not properly handled, which can lead to unhandled rejections and application crashes.

### Why This Matters

| Issue                  | Impact                          | Solution                  |
| ---------------------- | ------------------------------- | ------------------------- |
| ðŸ’¥ **Crashes**         | Unhandled rejection crashes app | Add .catch() or try-catch |
| ðŸ™ˆ **Silent Failures** | Errors go unnoticed             | Handle all rejections     |
| ðŸ“Š **Monitoring**      | No visibility into async errors | Proper error reporting    |
| ðŸ”’ **Reliability**     | Unpredictable behavior          | Explicit error handling   |

## Examples

### âŒ Incorrect

```typescript
// Floating promise without handling
fetchData();

// Promise in expression without catch
someAsyncFunction().then(process);

// Async operation without await or catch
async function example() {
  riskyOperation(); // Floating promise
}

// Array of promises without proper handling
urls.map((url) => fetch(url)); // Results ignored
```

### âœ… Correct

```typescript
// Await with try-catch
async function example() {
  try {
    await fetchData();
  } catch (error) {
    console.error('Fetch failed:', error);
  }
}

// Promise with catch
fetchData()
  .then(process)
  .catch((error) => console.error(error));

// Explicit void for fire-and-forget (if intentional)
void logAnalytics(event);

// Promise.all with error handling
try {
  const results = await Promise.all(urls.map((url) => fetch(url)));
} catch (error) {
  console.error('One or more fetches failed:', error);
}

// Promise.allSettled for partial failure handling
const results = await Promise.allSettled(urls.map((url) => fetch(url)));
const failures = results.filter((r) => r.status === 'rejected');
```

## Configuration Examples

### Basic Usage

```javascript
{
  rules: {
    'maintainability/no-unhandled-promise': 'error'
  }
}
```

## Related Rules

- [`error-message`](./error-message.mdx) - Meaningful error messages
- [`no-silent-errors`](./no-silent-errors.mdx) - Prevent ignoring errors
- [`no-missing-error-context`](./no-missing-error-context.mdx) - Error context preservation

## Further Reading

- **[Promise Error Handling](https://javascript.info/promise-error-handling)** - JavaScript.info guide
- **[Async/Await Error Handling](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Promises)** - MDN documentation
- **[ESLint MCP Setup](https://eslint.org/docs/latest/use/mcp)** - Enable AI assistant integration
